File: leetcode/decode_ways/DecodeWays.cpp
#include <iostream> 
#include <vector> 

using namespace std; 

class Solution {
public:
    int numDecodings(string s) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        if (s.size() == 0) {
            return 0; 
        }
        
        vector<int> ways(s.size() + 1, 1);
        
        for (int i = s.size() - 1; i >= 0; i--) {
            if (s[i] == '0') {
                ways[i] = 0; 
            }
            else {
                ways[i] = ways[i + 1]; 
                
                if (i < (s.size() - 1) && 
                    (s[i] == '1' || (s[i] == '2' && s[i + 1] <= '6'))
                ) {
                    ways[i] = ways[i + 1] + ways[i + 2]; 
                }
            }
        } 
        
        return ways[0]; 
    }
};

// Test driver 
int main() {
    Solution sln; 
    cout << sln.numDecodings("2") << endl; 
    return 0; 
}

File: leetcode4/DecodeWays.java
public class Solution {
    public int numDecodings(String s) {
        int len = s.length(); 
        
        if (len == 0) {
            return 0; 
        }
        
        int[] arr = new int[len];
        
        if (s.charAt(0) != '0') {
            arr[0] = 1; 
        }
        
        for (int i = 1; i < len; i++) {
            if (s.charAt(i) == '0') {
                if (s.charAt(i - 1) != '1' && s.charAt(i - 1) != '2') {
                    return 0; 
                }
                
                if (i > 1) {
                    arr[i] = arr[i - 2]; 
                } else {
                    arr[i] = 1; 
                }
                  
            } else {
                if ((s.charAt(i - 1) == '2' && s.charAt(i) > '6')  || s.charAt(i - 1) > '2' || s.charAt(i - 1) == '0') {
                arr[i] = arr[i - 1]; 
            } else {
                if (i > 1) {
                    arr[i] = arr[i - 1] + arr[i - 2]; 
                } else {
                    arr[i] = arr[i - 1] + 1; 
                }
            }
            }
        }
        
        return arr[len - 1]; 
    }
}
